### PR TITLE
Fully functional MLIR CUDA backend for mmperf

### PR DESCRIPTION
**Pull request by chudur-budur**
These are the changes I made:

1. The code is nicely integrated into the upstream version and runs with `MLIR CUDA` backend.
2. mmperf code can be compiled with separately built `llvm`, so that re-compilation of the `llvm-project` submodule can be avoided. It was needed as I already had `llvm` on my system and I was running out of disk space while compiling `llvm-project` submodule.
3. The code runs and produces correct results.
4. The performance is identical to `CUBLAS` and `OPENBLAS` backend code. For example, in the case of `24x64x96`, `MLIR-CUDA`, `OPENBLAS` and `CUBLAS` all give 0.02 GFLOP.
5. Removed custom build scripts and shell scripts.   
6. The current version gives `muMemHostRegister` error, the solution would be to move the `gpu.host_register` op outside from the mlir function and the loop in `main.cc`. Couldn't figure this out. But the current code gives correct `matmult` result.

### CODE REVIEW

File: benchmark_sizes/benchmark_all_sizes.txt [hunk 1]
```diff
@@ -10,10 +10,10 @@
 384x256x256
 480x512x16
 480x512x256
-1024x1024x1024
+#1024x1024x1024
```
Comment (powderluv) [inline benchmark_sizes/benchmark_all_sizes.txt:13]: why ?

File: benchmark_sizes/benchmark_all_sizes.txt [hunk 2]
```diff
@@ -10,10 +10,10 @@
...
 #1020x1152x1152
 #1920x2304x2304
-2304x2304x2560
+#2304x2304x2560
```
Comment (powderluv) [inline benchmark_sizes/benchmark_all_sizes.txt:16]: why do we comment it out ?

File: cmake/common.cmake [hunk 3]
```diff
@@ -45,11 +47,13 @@ set(VARS_TO_COPY
     USE_CUBLAS
     USE_HALIDE
     USE_MLIR
+    USE_MLIR_CUDA
     USE_MKL
     USE_NAIVE
     USE_OPENBLAS
     USE_RUY
     USE_TVM
+    USE_CUDA
```
Comment (powderluv) [inline cmake/common.cmake:56]: we probably want to make this USE_TVM_CUDA then

File: matmul/CMakeLists.txt [hunk 4]
```diff
@@ -63,7 +63,7 @@ if(USE_MLIR OR USE_NODAI)
       else()
         set(TILE_SIZES_LOCAL "${TILE_SIZES}")
         set(REGISTER_TILE_SIZES_LOCAL "${REGISTER_TILE_SIZES}")
-	set(COPY_FILL_TILE_SIZES_LOCAL "${COPY_FILL_TILE_SIZES}")
+	    set(COPY_FILL_TILE_SIZES_LOCAL "${COPY_FILL_TILE_SIZES}")
```
Comment (powderluv) [inline matmul/CMakeLists.txt:66]: whitespace ?

File: matmul/CMakeLists.txt [hunk 5]
```diff
@@ -116,8 +125,69 @@ if(USE_MLIR OR USE_NODAI)
     endif()
   endfunction()
 
+  set(TEMP ${LLVM_DIR})
```
Comment (powderluv) [inline matmul/CMakeLists.txt:128]: dont name variables TEMP. Is it getting over written in add_subdirectory ?

File: matmul/CMakeLists.txt [hunk 6]
```diff
@@ -116,8 +125,69 @@ if(USE_MLIR OR USE_NODAI)
...
   # Build matmul-compile and pass
   add_subdirectory(matmul-compile)
+  set(LLVM_DIR ${TEMP})
+endif()
+
+if(USE_MLIR_CUDA)
+    function(compile_mlir_cuda mlir_prefix M N K)
```
Comment (powderluv) [inline matmul/CMakeLists.txt:135]: isn't 90% of this function the same as the USE_MLIR case ?

File: matmul/CMakeLists.txt [hunk 7]
```diff
@@ -116,8 +125,69 @@ if(USE_MLIR OR USE_NODAI)
...
+          else()
+              if(NOT LLVM_DIR STREQUAL "")
+                  set(MLIR_OPT ${LLVM_DIR}/bin/mlir-opt)
+                  set(MLIR_TRANSLATE ${LLVM_DIR}/bin/mlir-translate)
+              else()
+                  message(FATAL_ERROR "LLVM_DIR not set, aborting." )
+              endif()
+          endif()
+          
+          add_custom_command(
```
Comment (powderluv) [inline matmul/CMakeLists.txt:180]: maybe do this only for USE_MLIR_CUDA and keep the above code common

File: matmul/compile.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    --lower-affine \
+    --convert-scf-to-std \
+    --canonicalize \
+    --pass-pipeline="gpu.module(strip-debuginfo, convert-gpu-to-nvvm, gpu-to-cubin)" \
+    --gpu-to-llvm 2>&1 >matmul-gpu.mlir.out 
+
+mlir-translate matmul-gpu.mlir.out --mlir-to-llvmir \
+    | opt -O3 -S | llc -O3 | as - -o matmul-gpu.mlir.o
+
+clang++-11 matmul-gpu.mlir.o -lcuda \
```
Comment (powderluv) [inline matmul/compile.sh:20]: can this whole file be made custom targets in cmake ? that way you can use the clang that cmake is using and mlir-opt from the right location etc.

File: mmperf.py [hunk 9]
```diff
@@ -28,6 +28,7 @@
               'blis': 'mediumspringgreen',
               'blasfeo': 'olivedrab',
               'cublas': 'chocolate',
+              'cuda': 'green',
```
Comment (powderluv) [inline mmperf.py:31]: mlir_cuda

File: matmul/CMakeLists.txt (hunk 5 above)
Comment (chudur-budur) [reply matmul/CMakeLists.txt:128]: Yes, how do I not this getting modified?

File: matmul/compile.sh (hunk 8 above)
Comment (chudur-budur) [reply matmul/compile.sh:20]: I meant to remove compile.sh, forgot to exclude from the commit. The current CMakeFiles does the exact same thing.

File: matmul/matmul-compile/CMakeLists.txt [hunk 10]
```diff
@@ -2,7 +2,9 @@ cmake_minimum_required(VERSION 3.18)
 project(matmul_compile CXX)
 
 set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to conform to")
+set(TEMP ${LLVM_DIR})
 find_package(MLIR REQUIRED CONFIG)
```
Comment (chudur-budur) [inline matmul/matmul-compile/CMakeLists.txt:6]: This line also modifies LLVM_DIR, is there any way I can make this not to touch LLVM_DIR?

File: matmul/CMakeLists.txt (hunk 4 above)
Comment (chudur-budur) [reply matmul/CMakeLists.txt:66]: I think the white space in the upstream needs to be removed, it's an extra indentation.

**Pull request closed without merging**